libraries/lib-wave-track/CachingSequence.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "Sequence.h"
+#include "SequenceInterface.h"
+#include "SequenceSampleCache.h"
+
+class WAVE_TRACK_API CachingSequence final : public SequenceInterface
```

Comment:
I don't like this solution to the invalidation problem.  Distinguishing abstract SequenceInterface, and making this decorator, seems to me, excessive complication for the problem.

I would rather, as I said, let the Sequence and its cache be closely cooperating classes in one translation unit, and just invalidate the cache in all mutating member functions of Sequence.  There would be little need to disturb other code outside of Sequence.

This commit does add a lazy creation-on-demand of the cache when getting from the sequence -- that part is good.

Putting sufficient invalidations into Sequence.cpp so that we can examine it in isolation and be confident of cache correctness -- that is good.

Still there might be the other problem of "eviction."  Reclaiming the memory resources of the cache when it's not good to occupy them.

The old caches were short-lived -- they disappeared after rendering a spectrogram, or after destrying a Mixer object.

The new caches are internal cursors of Sequences, and might be more numerous than before, as they become per-clip and not per-track, and are not local variables in appropriate scopes.

Sequence (and therefore WaveClip, and therefore WaveTrack) might therefore benefit by exposing an `InvalidateCaches` function that the application can decide to call in certain contexts.

True, it might seem like WaveTrack becomes a "leaky abstraction," but then it already also exposes functions to count and visit blocks for various other purposes, so adding this might not be too bad.

I would at least be sure to empty the caches of the duplicate tracks that are in the Undo stack -- both when creating Undo states, and also when duplicating tracks out of the Undo state.

Actually that might even require a pure virtual function in Track for cache invalidation.  See `TrackListRestorer` where the duplications happen.